lxd/device/gpu.go

```diff
@@ -175,3 +187,502 @@ func gpuSelected(device config.Device, gpu api.ResourcesGPUCard) bool {
+	// Generate the CDI specification
+	spec, err := cdi.GenerateSpec(cdiID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to generate CDI spec: %w", err)
+	}
+
+	var targetName string
+	cdiName := cdiID.Name
+	if len(cdiName) == 1 {
+		// In the official CDI specification, and iGPU is represented as "gpu<idx>"
+		// So we use the cdi.SimpleGPU constant to represent it like a classic GPU.
+		targetName = cdiName[0]
+	} else if len(cdiName) == 2 && cdiID.Class == cdi.MIG {
+		targetName = fmt.Sprintf("%s:%s", cdiName[0], cdiName[1])
+	} else {
+		return nil, nil, fmt.Errorf("invalid CDI locator %v", cdiName)
+	}
+
+	// Initialize the hooks as empty
+	hooks := &cdi.Hooks{ContainerRootFS: inst.RootfsPath()}
+	mounts := make([]*specs.Mount, 0)
+	nativeDevices := make(map[string]map[string]string)
+
+	// Translate the CDI specification into device run config items
+	// First, get the device specific configuration
+	for _, device := range spec.Devices {
+		if device.Name == targetName {
+			for _, d := range device.ContainerEdits.DeviceNodes {
+				err = constructCDIDevNodeList(nativeDevices, d)
+				if err != nil {
+					return nil, nil, err
+				}
+			}
+
+			for _, hook := range device.ContainerEdits.Hooks {
+				err = constructCDIHook(hook, hooks)
+				if err != nil {
+					return nil, nil, err
+				}
+			}
+
+			mounts = append(mounts, device.ContainerEdits.Mounts...)
+			break
+		}
+	}
+
+	// Then, get the common configuration
+	for _, generalDeviceNode := range spec.ContainerEdits.DeviceNodes {
+		err = constructCDIDevNodeList(nativeDevices, generalDeviceNode)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	for _, generalHook := range spec.ContainerEdits.Hooks {
+		err = constructCDIHook(generalHook, hooks)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	mounts = append(mounts, spec.ContainerEdits.Mounts...)
+
+	// CDI Mounts are converted to LXD `disk` devices
+	// If the instance is privileged, then we shift the disk device
+	instConf := inst.ExpandedConfig()
+	privileged := false
+	if instConf["security.privileged"] == "true" {
+		privileged = true
+	}
+
+	indirectSymlinks, err := constructCDIMountList(nativeDevices, cdiID, mounts, privileged)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// merge the indirectSymlinks to the list of symlinks to be create in the hooks
+	hooks.Symlinks = append(hooks.Symlinks, indirectSymlinks...)
+
+	return nativeDevices, hooks, nil
+}
+
+// startWithCDISpec will generate a CDI specification from a cdiID and translate its content
+// into disk and unix-char devices, start them and return a reference to these underlying devices.
+func startWithCDISpec(s *state.State, inst instance.Instance, cdiID cdi.ID, runConf *config.RunConfig) error {
+	nativeDevices, hooks, err := constructCDIDevices(inst, cdiID)
+	if err != nil {
+		return err
+	}
+
+	startCharDevFunc := func(charDev *unixCommon, rootRunConf *config.RunConfig) error {
+		charDevRunConf, err := charDev.Start()
```

Comment:
Please use the same method as the current gpu passthrough uses to pass through unix devices by using `unixDeviceSetupCharNum` rather than setting up a top-level LXD unix-char device.

Reply:
fixed